src/db/gc/mod.rs

```diff
@@ -0,0 +1,309 @@
+    pub async fn gc_loop(&mut self, interval: Duration) -> anyhow::Result<()> {
+        loop {
+            self.gc_workflow(interval).await?
+        }
+    }
+
+    // This function yields to the main GC loop if the conditions are not met for execution of the
+    // next step.
+    async fn gc_workflow(&mut self, interval: Duration) -> anyhow::Result<()> {
+        let depth = self.depth;
+        let tipset = (self.get_heaviest_tipset)();
+        let current_epoch = tipset.epoch();
+        // Don't run the GC if there aren't enough state-roots yet. Sleep and yield to the main loop
+        // in order to refresh the heaviest tipset value.
+        if depth > current_epoch {
+            time::sleep(self.block_time * (depth - current_epoch) as u32).await;
+            return anyhow::Ok(());
+        }
+
+        if self.marked.is_empty() {
+            info!("populate keys for GC");
+            self.populate()?;
+            self.epoch_marked = current_epoch;
+        }
+
+        let epochs_since_marked = current_epoch - self.epoch_marked;
+        // Don't proceed with next steps until we advance at least `depth` epochs. Sleep and yield
+        // to the main loop in order to refresh the heaviest tipset value.
+        if epochs_since_marked < depth {
+            time::sleep(self.block_time * (depth - epochs_since_marked) as u32).await;
+            return anyhow::Ok(());
+        }
+
+        info!("filter keys for GC");
+        self.filter(tipset, depth).await?;
+
+        info!("GC sweep");
+        self.sweep()?;
+
+        // Make sure we don't run the GC too often.
+        time::sleep(interval).await;
+
+        anyhow::Ok(())
+    }
+}
+#[cfg(test)]
+mod test {
+    use crate::blocks::{BlockHeader, Tipset};
+    use crate::chain::ChainStore;
+
+    use crate::db::{GarbageCollectable, MarkAndSweep, MemoryDB};
+    use crate::message_pool::test_provider::{mock_block, mock_block_with_parents};
+    use crate::networks::ChainConfig;
+
+    use crate::utils::db::CborStoreExt;
+
+    use core::time::Duration;
+
+    use crate::shim::clock::ChainEpoch;
+    use fvm_ipld_blockstore::Blockstore;
+    use std::sync::Arc;
+
+    fn insert_unrechable(db: impl Blockstore, quantity: u64) {
+        for idx in 0..quantity {
+            let block: BlockHeader = mock_block(1 + idx, 1 + quantity);
+            db.put_cbor_default(&block).unwrap();
+        }
+    }
+
+    fn run_to_epoch(db: impl Blockstore, cs: Arc<ChainStore<MemoryDB>>, epoch: ChainEpoch) {
+        let mut heaviest_tipset = cs.heaviest_tipset();
+
+        for _ in heaviest_tipset.epoch()..epoch {
+            let block2 = mock_block_with_parents(heaviest_tipset.as_ref(), 1, 1);
+            db.put_cbor_default(&block2).unwrap();
+
+            let tipset = Arc::new(Tipset::from(&block2));
+            cs.set_heaviest_tipset(tipset).unwrap();
+            heaviest_tipset = cs.heaviest_tipset();
+        }
+    }
+
+    #[tokio::test]
+    async fn test_populate() {
```

Comment:
Could you document this test? I'm unsure what it is testing.

Reply:
Sure.